### DISCUSSION TITLE
Feature: support for local-only CLI

### DISCUSSION THREAD

**Discussion by merkuriy** [cyrilletuzi/vscode-angular-schematics#430] [Ideas]
# Checks before posting an issue
<details><summary>RESULTS</summary>

- [x] I have read all the extension's documentation (README and docs/TROUBLESHOOTING)
- [x] My project meets the requirements indicated in README
- [x] I launched `ng g component hello` in the *VS Code* Terminal inside my project, and it works
- [x] I looked at the extension's logs ("Angular Schematics" Output channel, second tab left to the Terminal), and tried to fix my project's configuration accordingly
- [x] I understand I am writing to a *human being*, doing this open source project on his *free* time
- [x] I will be involved in my own issue, by answering questions and informing if the issue is resolved
</details>

# Configuration

- Ubuntu: 19.04
- VS Code: 1.46.1
- Angular CLI global (`ng version`): not installed
- Angular CLI local: 8.3.28
- Node: 12.11.1

# Description of the issue
AngularSchematics launch the global Angular CLI, not the local.

Related issues: #26

# How to reproduce the issue

AngularSchematics run: `ng g service shared/services/menu`
Result: `Command 'ng' not found, but can be installed with: sudo apt install ng-common`

Angular CLI Local is normal working:
```
./node_modules/.bin/ng g service shared/services/menu
npx ng g service shared/services/menu
yarn ng g service shared/services/menu
```

**Comment by cyrilletuzi** [comment]
This is working as intended: `@angular/cli` should be installed globally too, both as mentioned in the extension requirements and in [official Angular documentation](https://angular.io/guide/setup-local).

Note that the global installation is just to have direct access to the `ng` command in terminals, but then the CLI uses the local version when you run commands.

I'll look later if it is relevant to add back direct use of local CLI.

**Comment by cyrilletuzi** [comment]
So, there is already history about this.

In a previous version, I introduced the use of the local CLI directly (`./node_modules/.bin/ng g`), but as this command contains special characters, it resulted in issues in some specific shell installations (#27).

That is why I reverted and got back to the usage of the global CLI (`ng g`) in version 4.

However, before v4, Node `exec` function was used to launch commands. Since v4, the VS Code Terminal is used, so maybe the special characters issue would be gone.

But v4 introduced other features that makes such a feature less trivial to reimplement now: advanced workspaces are now supported (for example, `.bin/ng` could be in a parent directory).

So I think it is not worth it. Again:
- this is the recommandation of the official Angular documentation,
- it is really simple and quick to install the CLI globally,
- doing `ng g` directly still uses the local CLI.

Do you see a use case where installing the CLI globally is an issue?

If not, I will close this issue as working as intended. Someone can still do a Pull Request, but it should:
- not reintroduce #27 issue,
- work for all kind of workspaces.

**Comment by cyrilletuzi** [comment]
Closing this issue for house-keeping as there is no news. Feel free to answer and I will open it back.

**Comment by starikovs** [comment]
Hey guys!

I also would like to have local `ng` support. Probably instead of `./node_modules/.bin/ng g` the `npx ng g` can be used, which has the same behavior if you have `ng` installed locally and don't have it globally.

**Comment by cyrilletuzi** [comment]
Unfortunately it is not that easy:
- `npx` seems deprecated in npm v7, in favor of `npm exec`
- some users don't use npm but yarn instead

**Comment by cyrilletuzi** [comment]
I'll reopen this issue, but this feature needs a solution working with all configurations to be implemented.

**Comment by cyrilletuzi** [comment]
Hi @merkuriy, @DDeis, @ElodieC, @trubachoff and @starikovs, for information, an extended version of the extension is now available: **Schematics Pro**, in which a global Angular CLI installation is not required anymore. You will find all the details on the [dedicated website](https://www.cyrilletuzi.com/schematics-pro/).

To be exact, Schematics Pro totally avoids the issue by including its own schematics generator, so it is not dependent on Angular CLI at all. It also has the benefit to be faster and to avoid other issues (like shell issues, and dependencies not being in `node_modules`).

**Comment by cyrilletuzi** [comment]
Hi @merkuriy, @DDeis, @ElodieC, @trubachoff and @starikovs, just to let you know the latest version (v6) of the extension is not longer tied to the Angular CLI, so generation will work even without the Angular CLI installed globally.
